--- rate_control.c	2010-11-19 05:58:32.000000000 +0800
+++ rate_control.c89.c	2010-11-19 05:57:03.000000000 +0800
@@ -205,7 +205,7 @@
 	const bool		is_nonblocking
 	)
 {
-	int64_t new_rate_limit;
+	ssize_t new_rate_limit;
 
 /* pre-conditions */
 	pgm_assert (NULL != bucket);
@@ -215,6 +215,7 @@
 		return TRUE;
 
 	pgm_spinlock_lock (&bucket->spinlock);
+	{
 	pgm_time_t now = pgm_time_update_now();
 
 	if (bucket->rate_per_msec)
@@ -223,7 +224,9 @@
 		if (time_since_last_rate_check > pgm_msecs(1)) 
 			new_rate_limit = bucket->rate_per_msec;
 		else {
+#pragma warning( disable : 4244 )
 			new_rate_limit = bucket->rate_limit + ((bucket->rate_per_msec * time_since_last_rate_check) / 1000UL);
+#pragma warning( default : 4244 )
 			if (new_rate_limit > bucket->rate_per_msec)
 				new_rate_limit = bucket->rate_per_msec;
 		}
@@ -234,7 +237,9 @@
 		if (time_since_last_rate_check > pgm_secs(1)) 
 			new_rate_limit = bucket->rate_per_sec;
 		else {
+#pragma warning( disable : 4244 )
 			new_rate_limit = bucket->rate_limit + ((bucket->rate_per_sec * time_since_last_rate_check) / 1000000UL);
+#pragma warning( default : 4244 )
 			if (new_rate_limit > bucket->rate_per_sec)
 				new_rate_limit = bucket->rate_per_sec;
 		}
@@ -260,6 +265,7 @@
 	} 
 	pgm_spinlock_unlock (&bucket->spinlock);
 	return TRUE;
+	}
 }
 
 pgm_time_t
@@ -327,12 +333,14 @@
 		return 0;
 
 	pgm_spinlock_lock (&bucket->spinlock);
+	{
 	const pgm_time_t now = pgm_time_update_now();
 	const pgm_time_t time_since_last_rate_check = now - bucket->last_rate_check;
 	const int64_t bucket_bytes = bucket->rate_limit + pgm_to_secs (bucket->rate_per_sec * time_since_last_rate_check) - n;
 	pgm_spinlock_unlock (&bucket->spinlock);
 
 	return bucket_bytes >= 0 ? 0 : (bucket->rate_per_sec / -bucket_bytes);
+	}
 }
 
 /* eof */
