--- engine.c	2010-12-09 12:32:06.000000000 +0800
+++ engine.c89.c	2010-12-09 12:35:45.000000000 +0800
@@ -90,6 +90,7 @@
 	pgm_rand_init();
 
 #ifdef _WIN32
+	{
 	WORD wVersionRequested = MAKEWORD (2, 2);
 	WSADATA wsaData;
 	if (WSAStartup (wVersionRequested, &wsaData) != 0)
@@ -147,9 +148,11 @@
 		closesocket (sock);
 	}
 #	endif
+	}
 #endif /* _WIN32 */
 
 /* find PGM protocol id overriding default value, use first value from NIS */
+	{
 	const struct pgm_protoent_t *proto = pgm_getprotobyname ("pgm");
 	if (proto != NULL) {
 		if (proto->p_proto != pgm_ipproto_pgm) {
@@ -158,8 +161,10 @@
 			pgm_ipproto_pgm = proto->p_proto;
 		}
 	}
+	}
 
 /* ensure timing enabled */
+	{
 	pgm_error_t* sub_error = NULL;
 	if (!pgm_time_init (&sub_error)) {
 		if (sub_error)
@@ -169,9 +174,11 @@
 #endif
 		goto err_shutdown;
 	}
+	}
 
 /* receiver simulated loss rate */
 #ifdef PGM_DEBUG
+	{
 	char* env;
 	size_t envlen;
 
@@ -184,6 +191,7 @@
 		}
 		free (env);
 	}
+	}
 #endif
 
 /* create global sock list lock */
