--- indextoname.c	2010-05-21 11:35:23.000000000 +0800
+++ indextoname.c89	2010-08-05 14:12:26.000000000 +0800
@@ -39,13 +39,25 @@
 	return if_indextoname (ifindex, ifname);
 #else
 	pgm_return_val_if_fail (NULL != ifname, NULL);
+	{
 
-	MIB_IFROW ifRow = { .dwIndex = ifindex };
+	MIB_IFROW ifRow;
+	ifRow.dwIndex = ifindex;
+	{
 	const DWORD dwRetval = GetIfEntry (&ifRow);
 	if (NO_ERROR != dwRetval)
 		return NULL;
+#ifdef _MSC_VER
+	{
+	int i;
+	wcstombs_s (&i, ifname, IF_NAMESIZE, ifRow.wszName, _TRUNCATE);
+	}
+#else
 	strcpy (ifname, (char*)ifRow.wszName);
+#endif
 	return ifname;
+	}
+	}
 #endif /* _WIN32 */
 }
 
