--- indextoaddr.c	2011-03-12 10:12:29.000000000 +0800
+++ indextoaddr.c89.c	2011-03-12 10:39:57.000000000 +0800
@@ -45,7 +45,9 @@
 
 	if (0 == ifindex)		/* any interface or address */
 	{
+#pragma warning( disable : 4244 )
 		ifsa->sa_family = iffamily;
+#pragma warning( default : 4244 )
 		switch (iffamily) {
 		case AF_INET:
 			((struct sockaddr_in*)ifsa)->sin_addr.s_addr = INADDR_ANY;
@@ -62,6 +64,7 @@
 		return TRUE;
 	}
 
+	{
 	struct pgm_ifaddrs_t *ifap, *ifa;
 	if (!pgm_getifaddrs (&ifap, error)) {
 		pgm_prefix_error (error,
@@ -75,6 +78,7 @@
 		    ifa->ifa_addr->sa_family != iffamily)
 			continue;
 
+		{
 		const unsigned i = pgm_if_nametoindex (iffamily, ifa->ifa_name);
 		pgm_assert (0 != i);
 		if (i == ifindex)
@@ -85,6 +89,7 @@
 			pgm_freeifaddrs (ifap);
 			return TRUE;
 		}
+		}
 	}
 
 	pgm_set_error (error,
@@ -94,6 +99,7 @@
 		     ifindex);
 	pgm_freeifaddrs (ifap);
 	return FALSE;
+	}
 }
 
 /* eof */
