--- messages.c	2010-10-07 11:04:07.000000000 +0800
+++ messages.c89	2010-10-07 11:06:16.000000000 +0800
@@ -19,6 +19,9 @@
  * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
  */
 
+#ifdef _MSC_VER
+#	include <io.h>
+#endif
 #include <stdarg.h>
 #include <stdio.h>
 #include <unistd.h>
@@ -164,14 +167,22 @@
 	char tbuf[1024];
 
 	pgm_mutex_lock (&messages_mutex);
+	{
 	const int offset = pgm_snprintf_s (tbuf, sizeof (tbuf), _TRUNCATE, "%s: ", log_level_text (log_level));
 	pgm_vsnprintf_s (tbuf + offset, sizeof(tbuf) - offset, _TRUNCATE, format, args);
+	}
 	if (log_handler)
 		log_handler (log_level, tbuf, log_handler_closure);
 	else {
 /* ignore return value */
+#ifdef _MSC_VER
+		const int stdoutfd = _fileno (stdout);
+		_write (stdoutfd, tbuf, strlen (tbuf));
+		_write (stdoutfd, "\n", 1);
+#else
 		write (STDOUT_FILENO, tbuf, strlen (tbuf));
 		write (STDOUT_FILENO, "\n", 1);
+#endif
 	}
 		
 	pgm_mutex_unlock (&messages_mutex);
