--- nametoindex.c	2010-05-21 11:34:30.000000000 +0800
+++ nametoindex.c89	2010-08-04 12:57:20.000000000 +0800
@@ -78,6 +78,7 @@
 
 	pgm_assert (AF_INET6 != iffamily);
 
+	{
 	DWORD dwRet, ifIndex;
 	ULONG ulOutBufLen = DEFAULT_BUFFER_SIZE;
 	PIP_ADAPTER_INFO pAdapterInfo = NULL;
@@ -86,7 +87,9 @@
 /* loop to handle interfaces coming online causing a buffer overflow
  * between first call to list buffer length and second call to enumerate.
  */
-	for (unsigned i = MAX_TRIES; i; i--)
+	{
+	unsigned i;
+	for (i = MAX_TRIES; i; i--)
 	{
 		pgm_debug ("IP_ADAPTER_INFO buffer length %lu bytes.", ulOutBufLen);
 		pAdapterInfo = (IP_ADAPTER_INFO*)_pgm_heap_alloc (ulOutBufLen);
@@ -98,6 +101,7 @@
 			break;
 		}
 	}
+	}
 
 	switch (dwRet) {
 	case ERROR_SUCCESS:	/* NO_ERROR */
@@ -118,7 +122,9 @@
 		 pAdapter;
 		 pAdapter = pAdapter->Next)
 	{
-		for (IP_ADDR_STRING *pIPAddr = &pAdapter->IpAddressList;
+		{
+		IP_ADDR_STRING *pIPAddr;
+		for (pIPAddr = &pAdapter->IpAddressList;
 			 pIPAddr;
 			 pIPAddr = pIPAddr->Next)
 		{
@@ -132,11 +138,13 @@
 				return ifIndex;
 			}
 		}
+		}
 	}
 
 	if (pAdapterInfo)
 		_pgm_heap_free (pAdapterInfo);
 	return 0;
+	}
 }
 
 /* Retrieve adapter index via name.
@@ -155,6 +163,7 @@
 {
 	pgm_return_val_if_fail (NULL != ifname, 0);
 
+	{
 	ULONG ifIndex;
 	DWORD dwSize = DEFAULT_BUFFER_SIZE, dwRet;
 	IP_ADAPTER_ADDRESSES *pAdapterAddresses = NULL, *adapter;
@@ -170,7 +179,9 @@
 /* loop to handle interfaces coming online causing a buffer overflow
  * between first call to list buffer length and second call to enumerate.
  */
-	for (unsigned i = MAX_TRIES; i; i--)
+	{
+	unsigned i;
+	for (i = MAX_TRIES; i; i--)
 	{
 		pAdapterAddresses = (IP_ADAPTER_ADDRESSES*)_pgm_heap_alloc (dwSize);
 		dwRet = GetAdaptersAddresses (AF_UNSPEC,
@@ -188,6 +199,7 @@
 			break;
 		}
 	}
+	}
 
 	switch (dwRet) {
 	case ERROR_SUCCESS:
@@ -218,6 +230,7 @@
 	if (pAdapterAddresses)
 		_pgm_heap_free (pAdapterAddresses);
 	return 0;
+	}
 }
 #endif /* _WIN32 */
 
