--- net.c	2011-06-27 22:54:07.000000000 +0800
+++ net.c89.c	2011-10-06 01:37:13.000000000 +0800
@@ -67,6 +67,7 @@
 	pgm_assert( tolen > 0 );
 
 #ifdef NET_DEBUG
+	{
 	char saddr[INET_ADDRSTRLEN];
 	pgm_sockaddr_ntop (to, saddr, sizeof(saddr));
 	pgm_debug ("pgm_sendto (sock:%p use_rate_limit:%s minor_rate_control:%p use_router_alert:%s buf:%p len:%" PRIzu " to:%s [toport:%d] tolen:%d)",
@@ -75,12 +76,14 @@
 		(const void*)minor_rate_control,
 		use_router_alert ? "TRUE" : "FALSE",
 		(const void*)buf,
-		len,
+		(unsigned long)len,
 		saddr,
 		ntohs (((const struct sockaddr_in*)to)->sin_port),
 		(int)tolen);
+	}
 #endif
 
+	{
 	const SOCKET send_sock = use_router_alert ? sock->send_with_router_alert_sock : sock->send_sock;
 
 	if (use_rate_limit)
@@ -108,9 +111,11 @@
 	if (-1 != hops)
 		pgm_sockaddr_multicast_hops (send_sock, sock->send_gsr.gsr_group.ss_family, hops);
 
-	ssize_t sent = sendto (send_sock, buf, len, 0, to, (socklen_t)tolen);
-	pgm_debug ("sendto returned %" PRIzd, sent);
-	if (sent < 0) {
+	{
+	ssize_t sent = sendto (send_sock, buf, (int)len, 0, to, (socklen_t)tolen);
+	pgm_debug ("sendto returned %" PRIzd, (long)sent);
+	if (sent < 0)
+	{
 		int save_errno = pgm_get_last_sock_error();
 		if (PGM_UNLIKELY(save_errno != PGM_SOCK_ENETUNREACH &&	/* Network is unreachable */
 		 		 save_errno != PGM_SOCK_EHOSTUNREACH &&	/* No route to host */
@@ -126,23 +131,24 @@
 			const int ready = poll (&p, 1, 500 /* ms */);
 #else
 			fd_set writefds;
+			struct timeval tv;
+			int ready, n_fds;
+
 			FD_ZERO(&writefds);
 			FD_SET(send_sock, &writefds);
 #	ifndef _WIN32
-			const int n_fds = send_sock + 1;	/* largest fd + 1 */
+			n_fds = send_sock + 1;		/* largest fd + 1 */
 #	else
-			const int n_fds = 1;			/* count of fds */
+			n_fds = 1;			/* count of fds */
 #	endif
-			struct timeval tv = {
-				.tv_sec  = 0,
-				.tv_usec = 500 /* ms */ * 1000
-			};
-			const int ready = select (n_fds, NULL, &writefds, NULL, &tv);
+			tv.tv_sec  = 0;
+			tv.tv_usec = 500 /* ms */ * 1000;
+			ready = select (n_fds, NULL, &writefds, NULL, &tv);
 #endif /* HAVE_POLL */
 			if (ready > 0)
 			{
-				sent = sendto (send_sock, buf, len, 0, to, (socklen_t)tolen);
-				if ( sent < 0 )
+				sent = sendto (send_sock, buf, (int)len, 0, to, (socklen_t)tolen);
+				if (sent < 0)
 				{
 					char errbuf[1024];
 					char toaddr[INET6_ADDRSTRLEN];
@@ -174,7 +180,9 @@
 		pgm_sockaddr_multicast_hops (send_sock, sock->send_gsr.gsr_group.ss_family, sock->hops);
 	if (!use_router_alert && sock->can_send_data)
 		pgm_mutex_unlock (&sock->send_mutex);
-	return sent;
+	return (ssize_t)sent;
+	}
+	}
 }
 
 /* socket helper, for setting pipe ends non-blocking
