--- socket.c	2010-10-13 16:09:32.000000000 +0800
+++ socket.c89.c	2010-10-13 16:10:02.000000000 +0800
@@ -244,7 +244,9 @@
 	new_sock->adv_mode	= 0;	/* advance with time */
 
 /* PGMCC */
+#pragma warning( disable : 4244 )
 	new_sock->acker_nla.ss_family = family;
+#pragma warning( default : 4244 )
 
 /* source-side */
 	pgm_mutex_init (&new_sock->source_mutex);
@@ -324,6 +326,7 @@
 /* Stevens: "SO_REUSEADDR has datatype int."
  */
 		pgm_trace (PGM_LOG_ROLE_NETWORK,_("Set socket sharing."));
+		{
 		const int v = 1;
 		if (SOCKET_ERROR == setsockopt (new_sock->recv_sock, SOL_SOCKET, SO_REUSEADDR, (const char*)&v, sizeof(v)) ||
 		    SOCKET_ERROR == setsockopt (new_sock->send_sock, SOL_SOCKET, SO_REUSEADDR, (const char*)&v, sizeof(v)) ||
@@ -338,10 +341,12 @@
 				       pgm_sock_strerror_s (errbuf, sizeof (errbuf), save_errno));
 			goto err_destroy;
 		}
+		}
 
 /* request extra packet information to determine destination address on each packet */
 #ifndef CONFIG_TARGET_WINE
 		pgm_trace (PGM_LOG_ROLE_NETWORK,_("Request socket packet-info."));
+		{
 		const sa_family_t recv_family = new_sock->family;
 		if (SOCKET_ERROR == pgm_sockaddr_pktinfo (new_sock->recv_sock, recv_family, TRUE))
 		{
@@ -354,6 +359,7 @@
 				       pgm_sock_strerror_s (errbuf, sizeof (errbuf), save_errno));
 			goto err_destroy;
 		}
+		}
 #endif
 	}
 	else
@@ -610,8 +616,11 @@
 		{
 			int*restrict intervals = (int*restrict)optval;
 			*optlen = sock->spm_heartbeat_len;
-			for (unsigned i = 0; i < sock->spm_heartbeat_len; i++)
+			{
+			unsigned i;
+			for (i = 0; i < sock->spm_heartbeat_len; i++)
 				intervals[i] = sock->spm_heartbeat_interval[i + 1];
+			}
 		}
 		status = TRUE;
 		break;
@@ -1017,8 +1026,11 @@
 			sock->spm_heartbeat_len = optlen / sizeof (int);
 			sock->spm_heartbeat_interval = pgm_new (unsigned, sock->spm_heartbeat_len + 1);
 			sock->spm_heartbeat_interval[0] = 0;
-			for (unsigned i = 0; i < sock->spm_heartbeat_len; i++)
+			{
+			unsigned i;
+			for (i = 0; i < sock->spm_heartbeat_len; i++)
 				sock->spm_heartbeat_interval[i + 1] = ((const int*)optval)[i];
+			}
 		}
 		status = TRUE;
 		break;
@@ -1214,6 +1226,7 @@
 				break;
 			if (PGM_UNLIKELY(fecinfo->group_size > fecinfo->block_size))
 				break;
+			{
 			const uint8_t parity_packets = fecinfo->block_size - fecinfo->group_size;
 /* technically could re-send previous packets */
 			if (PGM_UNLIKELY(fecinfo->proactive_packets > parity_packets))
@@ -1230,6 +1243,7 @@
 			sock->rs_n			= fecinfo->block_size;
 			sock->rs_k			= fecinfo->group_size;
 			sock->rs_proactive_h		= fecinfo->proactive_packets;
+			}
 		}
 		status = TRUE;
 		break;
@@ -1353,7 +1367,9 @@
 		{
 			const struct group_req* gr = optval;
 /* verify not duplicate group/interface pairing */
-			for (unsigned i = 0; i < sock->recv_gsr_len; i++)
+			{
+			unsigned i;
+			for (i = 0; i < sock->recv_gsr_len; i++)
 			{
 				if (pgm_sockaddr_cmp ((const struct sockaddr*)&gr->gr_group, (struct sockaddr*)&sock->recv_gsr[i].gsr_group)  == 0 &&
 				    pgm_sockaddr_cmp ((const struct sockaddr*)&gr->gr_group, (struct sockaddr*)&sock->recv_gsr[i].gsr_source) == 0 &&
@@ -1372,6 +1388,7 @@
 					break;
 				}
 			}
+			}
 			if (PGM_UNLIKELY(sock->family != gr->gr_group.ss_family))
 				break;
 			sock->recv_gsr[sock->recv_gsr_len].gsr_interface = gr->gr_interface;
@@ -1403,7 +1420,9 @@
 			break;
 		{
 			const struct group_req* gr = optval;
-			for (unsigned i = 0; i < sock->recv_gsr_len;)
+			{
+			unsigned i;
+			for (i = 0; i < sock->recv_gsr_len;)
 			{
 				if ((pgm_sockaddr_cmp ((const struct sockaddr*)&gr->gr_group, (struct sockaddr*)&sock->recv_gsr[i].gsr_group) == 0) &&
 /* drop all matching receiver entries */
@@ -1420,6 +1439,7 @@
 				}
 				i++;
 			}
+			}
 			if (PGM_UNLIKELY(sock->family != gr->gr_group.ss_family))
 				break;
 			if (SOCKET_ERROR == pgm_sockaddr_leave_group (sock->recv_sock, sock->family, gr))
@@ -1478,7 +1498,9 @@
 		{
 			const struct group_source_req* gsr = optval;
 /* verify if existing group/interface pairing */
-			for (unsigned i = 0; i < sock->recv_gsr_len; i++)
+			{
+			unsigned i;
+			for (i = 0; i < sock->recv_gsr_len; i++)
 			{
 				if (pgm_sockaddr_cmp ((const struct sockaddr*)&gsr->gsr_group, (struct sockaddr*)&sock->recv_gsr[i].gsr_group) == 0 &&
 					(gsr->gsr_interface == sock->recv_gsr[i].gsr_interface ||
@@ -1503,6 +1525,7 @@
 					break;
 				}
 			}
+			}
 			if (PGM_UNLIKELY(sock->family != gsr->gsr_group.ss_family))
 				break;
 			if (PGM_UNLIKELY(sock->family != gsr->gsr_source.ss_family))
@@ -1525,7 +1548,9 @@
 		{
 			const struct group_source_req* gsr = optval;
 /* verify if existing group/interface pairing */
-			for (unsigned i = 0; i < sock->recv_gsr_len; i++)
+			{
+			unsigned i;
+			for (i = 0; i < sock->recv_gsr_len; i++)
 			{
 				if (pgm_sockaddr_cmp ((const struct sockaddr*)&gsr->gsr_group, (struct sockaddr*)&sock->recv_gsr[i].gsr_group)   == 0 &&
 				    pgm_sockaddr_cmp ((const struct sockaddr*)&gsr->gsr_source, (struct sockaddr*)&sock->recv_gsr[i].gsr_source) == 0 &&
@@ -1539,6 +1564,7 @@
 					}
 				}
 			}
+			}
 			if (PGM_UNLIKELY(sock->family != gsr->gsr_group.ss_family))
 				break;
 			if (PGM_UNLIKELY(sock->family != gsr->gsr_source.ss_family))
@@ -1842,7 +1868,7 @@
 
 /* determine IP header size for rate regulation engine & stats */
 	sock->iphdr_len = (AF_INET == sock->family) ? sizeof(struct pgm_ip) : sizeof(struct pgm_ip6_hdr);
-	pgm_trace (PGM_LOG_ROLE_NETWORK,"Assuming IP header size of %zu bytes", sock->iphdr_len);
+	pgm_trace (PGM_LOG_ROLE_NETWORK,"Assuming IP header size of %lu bytes", (unsigned long)sock->iphdr_len);
 
 	if (sock->udp_encap_ucast_port) {
 		const size_t udphdr_len = sizeof(struct pgm_udphdr);
@@ -1850,9 +1876,11 @@
 		sock->iphdr_len += udphdr_len;
 	}
 
+	{
 	const sa_family_t pgmcc_family = sock->use_pgmcc ? sock->family : 0;
 	sock->max_tsdu = sock->max_tpdu - sock->iphdr_len - pgm_pkt_offset (FALSE, pgmcc_family);
 	sock->max_tsdu_fragment = sock->max_tpdu - sock->iphdr_len - pgm_pkt_offset (TRUE, pgmcc_family);
+	{
 	const unsigned max_fragments = sock->txw_sqns ? MIN( PGM_MAX_FRAGMENTS, sock->txw_sqns ) : PGM_MAX_FRAGMENTS;
 	sock->max_apdu = MIN( PGM_MAX_APDU, max_fragments * sock->max_tsdu_fragment );
 
@@ -1898,6 +1926,7 @@
  */
 /* TODO: different ports requires a new bound socket */
 
+	{
 	union {
 		struct sockaddr		sa;
 		struct sockaddr_in	s4;
@@ -2066,6 +2095,7 @@
 
 /* save send side address for broadcasting as source nla */
 	memcpy (&sock->send_addr, &send_addr, pgm_sockaddr_len ((struct sockaddr*)&send_addr));
+	}
 
 /* rx to nak processor notify channel */
 	if (sock->can_send_data)
@@ -2073,13 +2103,8 @@
 /* setup rate control */
 		if (sock->txw_max_rte)
 		{
-#ifndef _MSC_VER
-			pgm_trace (PGM_LOG_ROLE_RATE_CONTROL,_("Setting rate regulation to %zd bytes per second."),
-					sock->txw_max_rte);
-#else
 			pgm_trace (PGM_LOG_ROLE_RATE_CONTROL,_("Setting rate regulation to %ld bytes per second."),
 					(long)sock->txw_max_rte);
-#endif
 	
 			pgm_rate_create (&sock->rate_control, sock->txw_max_rte, sock->iphdr_len, sock->max_tpdu);
 			sock->is_controlled_spm   = TRUE;	/* must always be set */
@@ -2104,6 +2129,8 @@
 	pgm_rwlock_writer_unlock (&sock->lock);
 	pgm_debug ("PGM socket successfully bound.");
 	return TRUE;
+	}
+	}
 }
 
 bool
@@ -2117,11 +2144,14 @@
 #ifdef CONFIG_TARGET_WINE
 	pgm_return_val_if_fail (sock->recv_gsr_len == 1, FALSE);
 #endif
-	for (unsigned i = 0; i < sock->recv_gsr_len; i++)
+	{
+	unsigned i;
+	for (i = 0; i < sock->recv_gsr_len; i++)
 	{
 		pgm_return_val_if_fail (sock->recv_gsr[i].gsr_group.ss_family == sock->recv_gsr[0].gsr_group.ss_family, FALSE);
 		pgm_return_val_if_fail (sock->recv_gsr[i].gsr_group.ss_family == sock->recv_gsr[i].gsr_source.ss_family, FALSE);
 	}
+	}
 	pgm_return_val_if_fail (sock->send_gsr.gsr_group.ss_family == sock->recv_gsr[0].gsr_group.ss_family, FALSE);
 /* shutdown */
 	if (PGM_UNLIKELY(!pgm_rwlock_writer_trylock (&sock->lock)))
@@ -2231,6 +2261,7 @@
 		return -1;
 	}
 
+	{
 	const bool is_congested = (sock->use_pgmcc && sock->tokens < pgm_fp8 (1)) ? TRUE : FALSE;
 
 	if (readfds)
@@ -2255,6 +2286,7 @@
 #endif
 			}
 		}
+		{
 		const SOCKET pending_fd = pgm_notify_get_fd (&sock->pending_notify);
 		FD_SET(pending_fd, readfds);
 #ifndef _WIN32
@@ -2262,6 +2294,7 @@
 #else
 		fds++;
 #endif
+		}
 	}
 
 	if (sock->can_send_data && writefds && !is_congested)
@@ -2279,6 +2312,7 @@
 #else
 	return *n_fds + fds;
 #endif
+	}
 }
 
 #ifdef CONFIG_HAVE_POLL
