--- tsi.c	2011-03-12 10:33:45.000000000 +0800
+++ tsi.c89.c	2011-03-12 11:13:49.000000000 +0800
@@ -46,11 +46,13 @@
 	pgm_return_val_if_fail (NULL != buf, -1);
 	pgm_return_val_if_fail (bufsize > 0, -1);
 
+	{
 	const uint8_t* gsi = (const uint8_t*)tsi;
 	const uint16_t source_port = tsi->sport;
 
 	return pgm_snprintf_s (buf, bufsize, _TRUNCATE, "%u.%u.%u.%u.%u.%u.%u",
 				 gsi[0], gsi[1], gsi[2], gsi[3], gsi[4], gsi[5], ntohs (source_port));
+	}
 }
 
 /* transform TSI to ASCII string form.
