--- time.c	2010-11-05 05:07:57.000000000 +0800
+++ time.c89.c	2010-11-05 05:07:48.000000000 +0800
@@ -320,6 +320,7 @@
 /* attempt to parse clock ticks from kernel
  */
 		FILE	*fp = fopen ("/proc/cpuinfo", "r");
+
 		if (fp)
 		{
 			char buffer[1024];
@@ -334,6 +335,7 @@
 			fclose (fp);
 		}
 #elif defined(_WIN32)
+		{
 		LARGE_INTEGER frequency;
 		if (QueryPerformanceFrequency (&frequency))
 		{
@@ -349,6 +351,7 @@
 				       _("No supported high-resolution performance counter: %s"),
 				       pgm_win_strerror (winstr, sizeof (winstr), save_errno));
 		}
+		}
 #endif /* !_WIN32 */
 
 /* e.g. export RDTSC_FREQUENCY=3200.000000
@@ -413,6 +416,7 @@
 
 /* update Windows timer resolution to 1ms */
 #ifdef _WIN32
+	{
 	TIMECAPS tc;
 	if (TIMERR_NOERROR == timeGetDevCaps (&tc, sizeof (TIMECAPS)))
 	{
@@ -424,6 +428,7 @@
 	{
 		pgm_warn (_("Unable to determine timer device resolution."));
 	}
+	}
 #endif
 
 	return TRUE;
@@ -778,11 +783,15 @@
 /* HPET counter tick period is in femto-seconds, a value of 0 is not permitted,
  * the value must be <= 0x05f5e100 or 100ns.
  */
+	{
 	const uint32_t hpet_period = *((uint32_t*)(hpet_ptr + HPET_COUNTER_CLK_PERIOD));
 	set_hpet_mul (hpet_period);
+	}
 #if defined( __x86_64__ ) || defined( __amd64 )
+	{
 	const uint32_t hpet_caps = *((uint32_t*)(hpet_ptr + HPET_GENERAL_CAPS_REGISTER));
 	hpet_wrap = hpet_caps & HPET_COUNT_SIZE_CAP ? 0 : (1ULL << 32);
+	}
 #else
 	hpet_wrap = 1ULL << 32;
 #endif
