--- rand.c	2011-06-27 22:55:20.000000000 +0800
+++ rand.c89.c	2011-10-06 01:39:17.000000000 +0800
@@ -35,7 +35,7 @@
 
 /* locals */
 
-static pgm_rand_t		global_rand = { .seed = 0 };
+static pgm_rand_t		global_rand = { 0 };
 static volatile uint32_t	rand_ref_count = 0;
 static pgm_mutex_t		rand_mutex;
 
@@ -87,8 +87,10 @@
 			return;
 	}
 #endif /* !_WIN32 */
+	{
 	const pgm_time_t now = pgm_time_update_now();
 	new_rand->seed = (uint32_t)pgm_to_msecs (now);
+	}
 }
 
 /* derived from POSIX.1-2001 example implementation of rand()
@@ -128,9 +130,11 @@
 	pgm_mutex_lock (&rand_mutex);
 	if (PGM_UNLIKELY(!global_rand.seed))
 		pgm_rand_create (&global_rand);
+	{
 	const uint32_t rand_value = pgm_rand_int (&global_rand);
 	pgm_mutex_unlock (&rand_mutex);
 	return rand_value;
+	}
 }
 
 PGM_GNUC_INTERNAL
